source/MaterialXContrib/MaterialXNode/MaterialXData.cpp

```diff
@@ -105,47 +105,65 @@ void MaterialXData::generateFragment(const mx::FileSearchPath& librarySearchPath
-        shaderRef ? mx::SPECULAR_ENVIRONMENT_FIS : mx::SPECULAR_ENVIRONMENT_NONE;
-
-    genContext.getOptions().hwMaxActiveLightSources = 0;
-    // For Maya we need to insert a V-flip fragment
-    genContext.getOptions().fileTextureVerticalFlip = true;
-
-    // Generate the fragment source (shader and XML wrapper).
-    _fragmentName = _element->getNamePath();
-    _fragmentName = mx::createValidName(_fragmentName);
-
-    _shader = shaderGenerator->generate(_fragmentName, _element, genContext);
-    if (!_shader)
     {
-        throw mx::Exception("Failed to generate shader");
+        mx::ShaderGeneratorPtr shaderGenerator = mx::GlslFragmentGenerator::create();
+        mx::GenContext genContext(shaderGenerator);
+
+        // Set up generator context. For shaders use FIS environment lookup,
+        // but disable this for textures to avoid additional unneeded XML parameter
+        // generation.
+        genContext.registerSourceCodeSearchPath(librarySearchPath);
+        genContext.getOptions().hwSpecularEnvironmentMethod =
+            shaderRef ? mx::SPECULAR_ENVIRONMENT_FIS : mx::SPECULAR_ENVIRONMENT_NONE;
+
+        genContext.getOptions().hwMaxActiveLightSources = 0;
+        // For Maya we need to insert a V-flip fragment
+        genContext.getOptions().fileTextureVerticalFlip = true;
+
+        // Generate the fragment source (shader and XML wrapper).
+        _fragmentName = _element->getNamePath();
+        _fragmentName = mx::createValidName(_fragmentName);
+
+        _shader = shaderGenerator->generate(_fragmentName, _element, genContext);
+        if (!_shader)
+        {
+            throw mx::Exception("Failed to generate shader");
+        }
     }
 
-    std::ostringstream stream;
-    mx::OgsXmlGenerator ogsXmlGenerator;
+    static const std::string FRAGMENT_NAME_TOKEN = "$fragmentName";
 
-    // Note: This name must match the the fragment name used for registration
-    // or the registration will fail.
-    ogsXmlGenerator.generate(_fragmentName, _shader.get(), nullptr, stream);
-    _fragmentSource = stream.str();
-    if (_fragmentSource.empty())
     {
-        throw mx::Exception("Generated shader source is empty");
+        std::ostringstream sourceStream;
+        mx::OgsXmlGenerator ogsXmlGenerator;
+
+        // Note: This name must match the the fragment name used for registration
+        // or the registration will fail.
+        ogsXmlGenerator.generate(FRAGMENT_NAME_TOKEN, _shader.get(), nullptr, sourceStream);
+        _fragmentSource = sourceStream.str();
+        if (_fragmentSource.empty())
+        {
+            throw mx::Exception("Generated shader source is empty");
+        }
     }
 
     // Strip out any '\r' characters.
     _fragmentSource.erase(
         std::remove(_fragmentSource.begin(), _fragmentSource.end(), '\r'), _fragmentSource.end()
     );
 
+    {
+        std::ostringstream nameStream;
+
+        const size_t sourceHash = std::hash<std::string>{}(_fragmentSource);
```

Comment:
Taking the hash of the generated code and appending it to the name making it unique for this unique combination of name and source code blob.